apisix/plugins/server-info.lua

```diff
@@ -0,0 +1,274 @@
+            default = 60,
+            minimum = 5,
+        }
+    }
+}
+
+local internal_status = ngx.shared.internal_status
+if not internal_status then
+    error("lua_shared_dict \"internal_status\" not configured")
+end
+
+
+local _M = {
+    version = 0.1,
+    priority = 990,
+    name = plugin_name,
+    schema = schema,
+}
+
+
+local function uninitialized_server_info()
+    return {
+        etcd_version     = "unknown",
+        hostname         = core.utils.gethostname(),
+        id               = core.id.get(),
+        version          = core.version.VERSION,
+        up_time          = ngx_time() - boot_time,
+        last_report_time = -1,
+    }
+end
+
+
+-- server information will be saved into shared memory only if the key
+-- "server_info" not exist if excl is true.
+local function save(data, excl)
+    local handler = excl and internal_status.add or internal_status.set
+
+    local ok, err = handler(internal_status, "server_info", data)
+    if not ok then
+        if excl and err == "exists" then
+            return true
+        end
+
+        return nil, err
+    end
+
+    return true
+end
+
+
+local function encode_and_save(server_info, excl)
+    local data, err = core.json.encode(server_info)
+    if not data then
+        return nil, err
+    end
+
+    return save(data, excl)
+end
+
+
+local function get()
+    local data, err = internal_status:get("server_info")
+    if err ~= nil then
+        return nil, err
+    end
+
+    if not data then
+        return uninitialized_server_info()
+    end
+
+    local server_info, err = core.json.decode(data)
+    if not server_info then
+        return nil, err
+    end
+
+    server_info.up_time = ngx_time() - boot_time
+    return server_info
+end
+
+
+local function get_server_info()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return 500, err
+    end
+
+    return 200, core.json.encode(server_info)
+end
+
+
+local function report()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return
+    end
+
+    if server_info.etcd_version == "unknown" then
+        local res, err = core.etcd.server_version()
+        if not res then
+            core.log.error("failed to fetch etcd version: ", err)
+            return
+
+        elseif type(res.body) ~= "table" then
+            core.log.error("failed to fetch etcd version: bad version info")
+            return
+
+        else
+            server_info.etcd_version = res.body.etcdcluster
+        end
+    end
+
+    server_info.last_report_time = ngx_time()
+
+    local data, err = core.json.encode(server_info)
+    if not data then
+        core.log.error("failed to encode server_info: ", err)
+        return
+    end
+
+    local key = "/data_plane/server_info/" .. server_info.id
+    local ok, err = core.etcd.set(key, data, 180)
```

Comment:
I think it would be better to let the CP clean up the data. For example, if the DP is just shutdowned for maintain, it would be better to keep the info data in etcd.

BTW, `180` is rather short compared with the 60 second report internal. And the number is undocumented. People will shout at us when they find their data is disappeared.

Reply:
> I think it would be better to let the CP clean up the data. For example, if the DP is just shutdowned for maintain, it would be better to keep the info data in etcd.

That can be solved by a long TTL.

> BTW, 180 is rather short compared with the 60 second report internal. And the number is undocumented. People will shout at us when they find their data is disappeared.

OK, we can add another item like `report_ttl` in the plugin_attr.